hasher.go

```diff
@@ -30,22 +32,66 @@ func newHasher(hasherProvider IonHasherProvider) *hasher {
 	return &hasher{hasherProvider, currentHasher, hasherStack}
 }
 
-func (h *hasher) scalar() {
-	panic("implement me")
+func (h *hasher) scalar(ionValue hashValue) {
+	h.currentHasher.scalar(ionValue)
 }
 
 func (h *hasher) stepIn(ionValue hashValue) {
-	panic("implement me")
+	var hashFunction IonHasher
+
+	_, isStructSerializer := h.currentHasher.(structSerializer)
+	if isStructSerializer {
+		hashFunction = h.hasherProvider.newHasher()
+	} else {
+		hashFunction = h.currentHasher.(scalarSerializer).hashFunction
+	}
+
+	if ionValue.ionType() == ion.StructType {
+		h.currentHasher = newStructSerializer(hashFunction, 0, h.hasherProvider)
+	} else {
+		h.currentHasher = newScalarSerializer(hashFunction, 0)
+	}
+
+	h.hasherStack.push(h.currentHasher)
+	h.currentHasher.stepIn(ionValue)
 }
 
-func (h *hasher) stepOut() {
-	panic("implement me")
+func (h *hasher) stepOut() error {
+	if h.depth() == 0 {
+		return &InvalidOperationError{"hasher", "stepOut", "Hasher cannot step out any further"}
+	}
+
+	h.currentHasher.stepOut()
+
+	poppedHasher, err := h.hasherStack.pop()
+	if err != nil {
+		return err
+	}
+	peekedHasher, err := h.hasherStack.peek()
+	if err != nil {
+		return err
+	}
+
+	h.currentHasher = peekedHasher.(serializer)
+
+	structHasher, isStructSerializer := h.currentHasher.(structSerializer)
+	if isStructSerializer {
+		digest := poppedHasher.(serializer).digest()
+		structHasher.appendFieldHash(digest)
+	}
+
+	return nil
 }
 
-func (h *hasher) digest() []byte {
-	panic("implement me")
+func (h *hasher) digest() ([]byte, error) {
+	if h.depth() == 0 {
+		return nil, &InvalidOperationError{
+			"hasher", "stepOut", "A digest may only be provided at the same depth hashing started"}
```

Comment:
This is not `stepOut` this is `digest`
I am not sure I get the condition either that leads to the error either. If the value is a single integer wouldn't that be at depth 0?

Reply:
Fixed stepOut to digest.
I also fixed the condition to h.depth() != 0.